Add a plugin to interact with the github status API.

shlevy opened the pull request:
Mutliple `<githubstatus>` sections are possible:
* jobs: regexp for jobs to match
* inputs: the input which corresponds to the github repo/rev whose
  status we want to report. Can be repeated
* authorization: Verbatim contents of the Authorization header. See
  https://developer.github.com/v3/#authentication.

shlevy commented:
@rickynils

Ericson2314 commented:
Excellent! This should bring us one step closer to building new PRs with hydra.nixos.org?

shlevy commented:
Yep. If this and #277 are merged, I have in mind how the last piece should look.

Ericson2314 commented:
:)

domenkozar commented:
This looks really useful - could we get some docs for the plugin? @edolstra should we start a section in hydra manual for plugins?

shlevy commented:
Happy to write up docs if @edolstra wants this merged

rickynils commented:
Nice work!

I might have missed something, but looking at the code, it seems the status context is hard-coded to `continuous-integration/hydra`?

So if we have multiple jobs running for the same set of inputs (which is the usual situation), will we not get multiple status reports with identical contexts (but with possibly different states) for the same repo/rev? I think the context actually should be the `project:jobset:job` id, possibly with some additional Hydra-specific prefix. Or the context could be configured manually in the `<githubstatus>` block, but then we would have to have one block per job.

That way, you could configure a GitHub protected branch to allow only commits that have passed a specific set of Hydra jobs. With the current implementation, you could only really decide if a commit has passed _any_ Hydra job, right?

shlevy commented:
Ah, I had in mind that the `jobs` regexp would be specific enough to ensure only a single build, but you're right that that makes more sense. Any reason to prefer `project:jobset:job` instead of `buildId`?

rickynils commented:
`buildId` is unique per build, not per job, isn't it? That wouldn't make sense for GitHub statuses, since it would change all the time. So you couldn't refer to it when you configure a protected branch in GitHub.

Using a specific `jobs` expression, the behavior I want could be encoded, but you would still be limited to a single status (`continuous-integration/hydra`). You wouldn't be able to have more fine-grained checks on the GitHub side. Maybe it wouldn't be such a severe limitation in practice, but it doesn't feel right.

An acceptable alternative to sending `project:jobset:job` would be to be able to configure the status name in the `<githubstatus>` block. You would need one block per status then, but I guess that is alright.

shlevy commented:
There doesn't seem to be any drawback to `project:jobset:job`, though if we don't have `buildId` then there might be a problem where an update of some other input that causes a rebuild (say, an update of nixpkgs that changes some dep of the project you're building) could cause multiple updates with the same context to the same commit.

rickynils commented:
Having `buildId` in the status context would make the status update useless, just saying :)

However, you are entirely correct that this feature doesn't blend well with multiple inputs. A Hydra job is usually not building/testing a single commit, it is building specific combinations of commits (from different repos). So, in the general case this plugin is probably not meaningful... And the status report should maybe be sent from a job-specific post-build hook instead. Which is being discussed in https://github.com/NixOS/hydra/pull/281

shlevy commented:
Hmm I'm still not clear on why the build ID would be useless. So the context changes from one build to the next, *within* the same build it will be the same (so first you'll see a 'pending' message, then a 'success' or 'failure' one). Can you explain more the workflow that requires the context be the same for all builds of the same job?

shlevy commented:
Oh, I guess it's the github protected branch feature, I see.

shlevy commented:
Well, if you're using protected branches, you probably want all the other inputs pegged at specific revisions anyway, right?

rickynils commented:
Yeah, that's right. And in that case this plugin would actually work just fine (if adjusted to send something job-constant as status context).

shlevy commented:
OK, so for manually reviewed PRs we want the build number in the context (so that we can see how the change fares against all possible input changes) and for automatic use cases like protected branches where we can't use a regexp to identify the relevant context we drop the build number. So does a config switch to go from `continuous-integration/hydra-project:jobset:job:build` to `continuous-integration/hydra-project:jobset:job` work for you?

rickynils commented:
@shlevy Yes, I think such a switch would do the trick.

(Btw, I've e-mailed you more details on my use case that I didn't have the energy to clean up for public consumption)

domenkozar commented:
@shlevy if you address the status context, I'll test this and report back.

shlevy commented:
Rebased and added an `excludeBuildFromContext` setting. Compare https://github.com/shlevy/hydra-github-status-test/pull/2 and https://github.com/shlevy/hydra-github-status-test/pull/3

shlevy commented:
@edolstra ping
@rickynils FYI

Ericson2314 commented:
That demo is mouthwatering!

domenkozar commented:
@shlevy can you post an example configuration (minus correct credentials)?

shlevy commented:
```
<githubstatus>
  jobs = test:pr:build
  inputs = src
  authorization = Basic notgivingyoumypasswordosorry
  excludeBuildFromContext = 1
</githubstatus>
```
`jobs` can be a regexp. `inputs` can be specified multiple times. This example matches only the `build` job of the `pr` jobset of the `test` project, sets a status associated with the revision of the `src` input when the build starts and when it finishes, and doesn't include the build's ID in the context of the status.

edolstra commented:
`excludeBuildFromContext` should be something like `display_build_id`. Otherwise the user has to guess what it means to exclude a build from a "context" (what context?).

shlevy commented:
`context` has a specific meaning for github statuses, and including the build ID should be the default IMO. And either way the build ID is displayed, it's just that it's not included in, well, the context :grin: . If you still think `display_build_id` is best I will switch to that.

robinp commented:
To clarify: `inputs` should be set to the jobset input name which triggers the source checkout (of type `Git checkout`). I was a bit confused, thinking it should be set to a path inside the repo.
